compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -1097,12 +1096,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
             // If the assignment expression itself is ill-formed, don't
             // bother emitting another error
-            if lhs_ty.references_error() || rhs_ty.references_error() {
+            let reported = if lhs_ty.references_error() || rhs_ty.references_error() {
                 err.delay_as_bug()
             } else {
-                err.emit();
-            }
-            return self.tcx.ty_error();
+                err.emit()
```

Comment:
`err.emit_unless` can be used here :thinking:

Reply:
change to 
```rust
 let reported = err.emit_unless(lhs_ty.references_error() || rhs_ty.references_error());
 return self.tcx.ty_error_with_guaranteed(reported);
```